### PR TITLE
Eupay 452 vrp consent event subscription

### PR DESCRIPTION
**Pull request by vivek0739**
## Context

To Improve the VRP experience, we want to get notified when User cancelled the VRP consent through ASPSP.
Open Banking provide Event API through which ASPSP can notify us whenever User revoked the consent. 

## Changes

-  Added event-openApi.yaml spec.
-  Implemented `subscribeToAnEvent`, `getAllEventResources`, `changeAnEventResource`, `deleteAnEventResource`
-  Unit test for each method.

### CODE REVIEW

File: src/main/java/com/transferwise/openbanking/client/api/event/EventClient.java [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+public interface EventClient {
+
+    /**
+     * @param eventSubscriptionRequest  request for event contains {callback URL, event type, version}
+     * @param aspspDetails -> detail of ASPSP which has the event to subscribe.
+     * @param softwareStatementDetails ->  The details of the software statement that the ASPSP registration uses
+     * @return OBEventSubscriptionResponse1 -> Event {subscription id, callback URL, version, event types}
+     */
+    OBEventSubscriptionResponse1 subscribeToAnEvent(OBEventSubscription1 eventSubscriptionRequest, AspspDetails aspspDetails,
+        SoftwareStatementDetails softwareStatementDetails);
```
Comment (ulianaostapenko) [inline src/main/java/com/transferwise/openbanking/client/api/event/EventClient.java:21]: As it doesn't fit in one line I would move each method param on a separate line so it is easier to read. 
```suggestion
    OBEventSubscriptionResponse1 subscribeToAnEvent(
        OBEventSubscription1 eventSubscriptionRequest,
        AspspDetails aspspDetails,
        SoftwareStatementDetails softwareStatementDetails);
```

File: src/main/java/com/transferwise/openbanking/client/api/event/RestEventClient.java [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+import org.springframework.web.client.RestOperations;
+
+public class RestEventClient extends BasePaymentClient implements EventClient {
+
+    private static final String BASE_ENDPOINT_PATH_FORMAT = "%s/open-banking/v3.%s/%s";
+    private static final String EVENT_SUBSCRIPTION_RESOURCE = "event-subscriptions";
+
+    private final JwtClaimsSigner jwtClaimsSigner;
+
+    protected RestEventClient(RestOperations restOperations,
```
Comment (ulianaostapenko) [inline src/main/java/com/transferwise/openbanking/client/api/event/RestEventClient.java:27]:
```suggestion
    protected RestEventClient(
        RestOperations restOperations,
```

File: src/main/java/com/transferwise/openbanking/client/api/event/RestEventClient.java [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+        OAuthClient oAuthClient,
+        JwtClaimsSigner jwtClaimsSigner
+    ) {
+        super(restOperations, jsonConverter, oAuthClient);
+        this.jwtClaimsSigner = jwtClaimsSigner;
+    }
+
+    @Override
+    public OBEventSubscriptionResponse1 subscribeToAnEvent(
+        OBEventSubscription1 eventSubscriptionRequest, AspspDetails aspspDetails,
```
Comment (ulianaostapenko) [inline src/main/java/com/transferwise/openbanking/client/api/event/RestEventClient.java:38]:
```suggestion
        OBEventSubscription1 eventSubscriptionRequest,
        AspspDetails aspspDetails,
```

File: src/main/java/com/transferwise/openbanking/client/api/event/RestEventClient.java [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+                String.class
+            );
+        } catch (RestClientResponseException e) {
+            OBErrorResponse1 errorResponse = mapBodyToObErrorResponse(e.getResponseBodyAsString());
+            throw new EventApiCallException("Call to subscribe event endpoint failed, body returned '" + e.getResponseBodyAsString() + "'", e, errorResponse);
+        } catch (RestClientException e) {
+            throw new EventApiCallException("Call to subscribe event endpoint failed, and no response body returned", e);
+        }
+        OBEventSubscriptionResponse1 eventSubscriptionResponse = jsonConverter.readValue(response.getBody(), OBEventSubscriptionResponse1.class);
+        return eventSubscriptionResponse;
```
Comment (ulianaostapenko) [inline src/main/java/com/transferwise/openbanking/client/api/event/RestEventClient.java:66]:
```suggestion
        return jsonConverter.readValue(response.getBody(), OBEventSubscriptionResponse1.class);
```

**Review by ulianaostapenko — CHANGES_REQUESTED** [review]

**Comment by ulianaostapenko** [comment]
Don't forget to bump version in gradle properties [here](https://github.com/transferwise/openbanking-client/blob/master/gradle.properties)

File: specs/events-openapi.yaml [hunk 5]
```diff
@@ -0,0 +1,765 @@
...
+        LastAvailableDateTime:
+          $ref: "#/components/schemas/ISODateTime"
+      additionalProperties: false
+    OBError1:
+      type: "object"
+      properties:
+        ErrorCode:
+          description: "Low level textual error code, e.g., UK.OBIE.Field.Missing"
+          type: "string"
+          x-namespaced-enum:
```
[the remark this answers is not in the document]
Comment (ulianaostapenko) [reply specs/events-openapi.yaml:541]: `ErrorCode` enum is not generated. Please resolve comments only if you applied the change or left a comment why you think this shouldn't be done

File: src/main/java/com/transferwise/openbanking/client/api/event/EventClient.java [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+
+    /**
+     * Create an Event Subscription :
+     * Asking a ASPSP to notify callback whenever there are updates on the given event.
+     *
+     * @param eventSubscriptionRequest  request for event contains {callback URL, event type, version}
+     * @param aspspDetails -> detail of ASPSP which has the event to subscribe.
+     * @param softwareStatementDetails ->  The details of the software statement that the ASPSP registration uses
+     * @return OBEventSubscriptionResponse1 -> Event {subscription id, callback URL, version, event types}
+     */
```
Comment (ulianaostapenko) [inline src/main/java/com/transferwise/openbanking/client/api/event/EventClient.java:22]: Please follow java documentation guidelines like ` @param nameOfParam<tab>short description what it is ` - no `{`, `}` or `->`. You can check how it is written in other clients

File: specs/events-openapi.yaml (hunk 5 above)
[the remark this answers is not in the document]
Comment (vivek0739) [reply specs/events-openapi.yaml:541]: I may have resolved it by mistake. I have made the changes for other comment in this https://github.com/transferwise/openbanking-client/commit/f7eab7883dc8fc7683b0219efa02ab818716df2f.
 
 @jonatwise helped me find this issue page: [#13 x-namespaced-enum not generating correctly when using Swagger-CodeGen](https://github.com/OpenBankingUK/read-write-api-specs/issues/13). It says that x-namespaced-enum  is an  open banking term. In java, It is a normal string and bank can send a customised string.

File: src/main/java/com/transferwise/openbanking/client/api/event/EventClient.java (hunk 6 above)
Comment (vivek0739) [reply src/main/java/com/transferwise/openbanking/client/api/event/EventClient.java]: Fixed in [c5c5c21](https://github.com/transferwise/openbanking-client/pull/170/commits/c5c5c21782dd2bc3e4d8cd3ffeacbf56dfa329cc).

**Comment by vivek0739** [comment]
> Don't forget to bump version in gradle properties [here](https://github.com/transferwise/openbanking-client/blob/master/gradle.properties)

**Comment by vivek0739** [comment]
Updated here: [c8bf287](https://github.com/transferwise/openbanking-client/pull/170/commits/c8bf287dc381439ae5446bfd1fe6bbcd7bcfd865)

File: specs/events-openapi.yaml (hunk 5 above)
[the remark this answers is not in the document]
Comment (ulianaostapenko) [reply specs/events-openapi.yaml:541]: Can you explain why do you prefer string here? 
I appreciate that link and I am aware of that. Banks can extend it in theory but in practice they don't do that. If we decide to have string I encourage you to change it for VRP and PISP packages so we have same style across all openbanking-clients API classes

File: src/main/java/com/transferwise/openbanking/client/api/event/RestEventClient.java [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+            OBErrorResponse1 errorResponse = mapBodyToObErrorResponse(e.getResponseBodyAsString());
+            throw new EventApiCallException("Call to subscribe event endpoint failed, body returned '" + e.getResponseBodyAsString() + "'", e, errorResponse);
+        } catch (RestClientException e) {
+            throw new EventApiCallException("Call to subscribe event endpoint failed, and no response body returned", e);
+        }
+        return jsonConverter.readValue(response.getBody(), OBEventSubscriptionResponse1.class);
+    }
+
+    @Override
+    public OBEventSubscriptionsResponse1 getAllEventResources( AspspDetails aspspDetails)  {
```
Comment (ulianaostapenko) [inline src/main/java/com/transferwise/openbanking/client/api/event/RestEventClient.java:74]:
```suggestion
    public OBEventSubscriptionsResponse1 getAllEventResources(AspspDetails aspspDetails)  {
```

File: src/test/java/com/transferwise/openbanking/client/api/event/RestEventClientTest.java [hunk 8]
```diff
@@ -0,0 +1,213 @@
...
+            .andRespond(
+                MockRestResponseCreators.withSuccess(jsonResponse, MediaType.APPLICATION_JSON));
+
+        var response = restEventClient.subscribeToAnEvent(
+            eventSubscriptionRequest, aspspDetails, softwareStatementDetails);
+        Assertions.assertEquals(mockEventSubscriptionResponse, response);
+        mockAspspServer.verify();
+    }
+
+    private OBEventSubscription1 aEventSubscriptionRequest() {
```
Comment (ulianaostapenko) [inline src/test/java/com/transferwise/openbanking/client/api/event/RestEventClientTest.java:131]: Please move all private methods after all public, default, protected methods.

**Review by ulianaostapenko — CHANGES_REQUESTED** [review]
Please update CHANGELOG file

**Review by ulianaostapenko — COMMENTED** [review]
Please squash all your commits before I can approve it as we cannot force push to master and you have 17 commits in the PR.
Also please rebase/merge master to get latest changes

**Comment by vivek0739** [comment]
> Please squash all your commits before I can approve it as we cannot force push to master and you have 17 commits in the PR. 
Did you mean squash and merge? which I will do when merging.  I didn't understand. You want to squash all these changes into one commit in this branch [EUPAY-452-vrp-consent-event-subscription](https://github.com/transferwise/openbanking-client/tree/EUPAY-452-vrp-consent-event-subscription) itself. 
We will all commits informations. 

> Also please rebase/merge master to get latest changes
I have rebased to master.

File: src/test/java/com/transferwise/openbanking/client/api/event/RestEventClientTest.java [hunk 9]
```diff
@@ -0,0 +1,215 @@
...
+import org.springframework.test.web.client.response.MockRestResponseCreators;
+import org.springframework.web.client.RestTemplate;
+
+@ExtendWith(MockitoExtension.class)
+@SuppressWarnings({
+    "PMD.AvoidDuplicateLiterals",
+    "checkstyle:membername",
+    "checkstyle:variabledeclarationusagedistance",
+    "checkstyle:methodname",
+    "checkstyle:abbreviationaswordinname"})
```
Comment (ulianaostapenko) [inline src/test/java/com/transferwise/openbanking/client/api/event/RestEventClientTest.java:45]: As you are adding this test can you fix "checkstyle:variabledeclarationusagedistance" to not mute it?

**Comment by ulianaostapenko** [comment]
Please make sure you answered and resolved all my questions in this PR.

> Did you mean squash and merge? which I will do when merging. I didn't understand. You want to squash all these changes into one commit in this branch [EUPAY-452-vrp-consent-event-subscription](https://github.com/transferwise/openbanking-client/tree/EUPAY-452-vrp-consent-event-subscription) itself.
We will all commits informations.

I mean squash them in the brach as you can make a mistake and choose `merge` instead `squash & merge` so all this commits will go to master. It has happen before, we cannot rewrite history on master and it is public library I advice you to squash them before. You can still have all your commit messaged listed in the commit.

File: specs/events-openapi.yaml (hunk 5 above)
[the remark this answers is not in the document]
Comment (vivek0739) [reply specs/events-openapi.yaml]: Changed in [7dcd03c](https://github.com/transferwise/openbanking-client/pull/170/commits/7dcd03cf287a7673791e0b1479fa520cc3e629d7)

File: src/test/java/com/transferwise/openbanking/client/api/event/RestEventClientTest.java (hunk 9 above)
Comment (vivek0739) [reply src/test/java/com/transferwise/openbanking/client/api/event/RestEventClientTest.java]: Removed in [7ac1674](https://github.com/transferwise/openbanking-client/pull/170/commits/7ac1674d34c7a2e6229e97a60602fe25f6e2302a).

**Review by ulianaostapenko — APPROVED** [review]
LGTM

**Review by ulianaostapenko — APPROVED** [review]

**Comment by vivek0739** [comment]
Merged with Master!
Description is branch name. We have just added support event subscription API. That's all we need in Open Banking client.

**Pull request merged**